tests/testthat/test-interpolate.R

```diff
@@ -22,38 +25,80 @@ testthat::test_that("`interpolate` works", {
+
+# new input
+dt <- data.table::data.table(
+  group = c(rep("a", 5), rep("b", 5)),
+  age = rep(c(1:5), 2),
+  outcome = c(seq(10, 50, 10), 9, 22, 27, 44, 51)
+)
+
+# setup expected output for linear `extrapolate`
+expected_output <- data.table::data.table(
+  group = c(rep("a", 8), rep("b", 8)),
+  age = rep(c(1:8), 2),
+  outcome = c(seq(10, 80, 10), 9, 22, 27, 44, 51, 62, 72.4, 82.8)
+)
+
+testthat::test_that("`extrapolate` works", {
 
   output <- testthat::expect_silent(
-    extrapolate(dt, id_cols = c("group", "x"), extrapolate_col = "x",
-                value_col = "y", extrapolate_vals = c(1:10),
-                method = "linear", n_groups_fit = 3)
+    extrapolate(dt, id_cols = c("group", "age"), extrapolate_col = "age",
+                value_col = "outcome", extrapolate_vals = c(1:8),
+                method = "linear", n_groups_fit = 4)
   )
   testthat::expect_equal(output, expected_output)
 
   output_roc <- testthat::expect_silent(
-    extrapolate(dt, id_cols = c("group", "x"), extrapolate_col = "x",
-                value_col = "y", extrapolate_vals = c(1:10),
-                method = "rate_of_change", n_groups_fit = 3)
+    extrapolate(dt, id_cols = c("group", "age"), extrapolate_col = "age",
+                value_col = "outcome", extrapolate_vals = c(1:8),
+                method = "rate_of_change", n_groups_fit = 4)
   )
   expected_output_roc <- copy(expected_output)
-  expected_output_roc$y <- rep(c(10, 20, 30, 40, 50, 64.5, 83.3, 107.6, 138.9, 179.3), 2)
-  testthat::expect_equivalent(output_roc, expected_output_roc, tolerance = 0.2)
+  expected_output_roc$outcome <- c(10, 20, 30, 40, 50, 67.86, 92.10, 125,
+                                   9, 22, 27, 33, 51, 67.50, 89.33, 118.23)
+  testthat::expect_equivalent(output_roc, expected_output_roc, tolerance = 0.1)
 
   output_unif <- testthat::expect_silent(
-    extrapolate(dt, id_cols = c("group", "x"), extrapolate_col = "x",
-                value_col = "y", extrapolate_vals = c(1:10),
-                method = "uniform", n_groups_fit = 3)
+    extrapolate(dt, id_cols = c("group", "age"), extrapolate_col = "age",
+                value_col = "outcome", extrapolate_vals = c(1:8),
+                method = "uniform", n_groups_fit = 4)
   )
 
   expected_output_unif <- copy(expected_output)
-  expected_output_unif[x > 5, y := 40]
+  expected_output_unif[age > 5 & group == "a", outcome := 35]
+  expected_output_unif[age > 5 & group == "b", outcome := 36]
   testthat::expect_equal(output_unif, expected_output_unif)
 
 })
+
+testthat::test_that("`extrapolate` works with id_cols that aren't present in every combination", {
+
+  dt <- data.table::data.table(
+    age_start = c(1:5),
+    year = 1,
+    outcome = seq(10, 50, 10)
+  )
+  dt[, age_end := age_start + 1]
+  for (i in 2:3) {
+    temp <- copy(dt[year == 1])
+    temp[, year := i]
+    temp[, outcome := outcome * year]
+    dt <- rbind(dt, temp)
+  }
```

Comment:
Isn't this just creating the same dataset for each year with a different outcome value?

Reply:
Could you just take `dt` from the previous test drop some of the rows for group "b"?

Reply:
Probably could do that, but I like having an example with `age_start` and `age_end`, and it's still a fairly small example, so going to keep for now.

Reply:
Is this actually creating a dataset with "id_cols that aren't present in every combination" though?

Reply:
Yes. The reason I identified this issue was because I had both `age_start` and `age_end` in `id_cols` and the function was creating all combinations of those variables. So you'd get age groups like 0-1, 0-2, 0-3, 0-4, 0-5, 1-0, 1-1, 1-2, ... etc. when really you just want age groups that are 1-year in width.